状态机+回调，

suspend为标识，这样在编译器的时候，会自动去做代码处理，完成状态机+回调的实现，当遇到内部调用挂起方法时，则会增加一个状态label，将后续的代码放到下一个label判断中等待调用

先有状态，回调与平时线程切换访问网络，传入callback，网络请求结束，切换线程调用callback处理结果的作用是一样的，区别在于，用状态机的方式，将异步代码用同步的方式完成。

```kotlin
suspend fun test(): Int{
    return 11
}

fun main(){

    GlobalScope.launch(Dispatchers.Main) {
        val aaa = "121"

        val bbb = withContext(Dispatchers.IO){
            return@withContext test()
        }

        print( aaa + bbb)
    }

}
```

反编译后代码

```java
public final class CoroutinesTestKt {
   @Nullable
   public static final Object test(@NotNull Continuation $completion) {
      return Boxing.boxInt(11);
   }

   public static final void main() {
      BuildersKt.launch$default((CoroutineScope)GlobalScope.INSTANCE, (CoroutineContext)Dispatchers.getMain(), (CoroutineStart)null, (Function2)(new Function2((Continuation)null) {
         Object L$0;
         int label; //初始值默认为0

         @Nullable
         public final Object invokeSuspend(@NotNull Object $result) {
            //标识为挂起
            Object var6 = IntrinsicsKt.getCOROUTINE_SUSPENDED();
            Object var10000;
            String aaa;
            switch(this.label) {
            case 0:
               ResultKt.throwOnFailure($result);
               aaa = "121";
               CoroutineContext var7 = (CoroutineContext)Dispatchers.getIO();
               Function2 var10001 = (Function2)(new Function2((Continuation)null) {
                  int label;

                  @Nullable
                  public final Object invokeSuspend(@NotNull Object $result) {
                     Object var2 = IntrinsicsKt.getCOROUTINE_SUSPENDED();
                     Object var10000;
                     switch(this.label) {
                     case 0:
                        ResultKt.throwOnFailure($result);
                        this.label = 1;
                        var10000 = CoroutinesTestKt.test(this);
                         //此时var10000已经获取到真实的值，!=var2，无需再挂起，直接返回var10000
                         //上层拿到var10000，发现不是挂起，则切换回原来的线程
                         //执行之前传进来的callback.invokeSuspends(result)将var10000传出去
                        if (var10000 == var2) {
                           return var2;
                        }
                        break;
                     case 1:
                        ResultKt.throwOnFailure($result);
                        var10000 = $result;
                        break;
                     default:
                        throw new IllegalStateException("call to 'resume' before 'invoke' with coroutine");
                     }

                     return var10000;
                  }

                  @NotNull
                  public final Continuation create(@Nullable Object value, @NotNull Continuation completion) {
                     Intrinsics.checkNotNullParameter(completion, "completion");
                     Function2 var3 = new <anonymous constructor>(completion);
                     return var3;
                  }

                  public final Object invoke(Object var1, Object var2) {
                     return ((<undefinedtype>)this.create(var1, (Continuation)var2)).invokeSuspend(Unit.INSTANCE);
                  }
               });
               this.L$0 = aaa;
               this.label = 1;
               //这里会去切换线程，将this(callback)传入，
               //子线程完成任务后切换原来线程并调用this.invokeSuspends(result)执行label == 1逻辑
               var10000 = BuildersKt.withContext(var7, var10001, this);
               //如果需要切换线程，则var10000会==var6，此时为挂起
               //需要等待下一次invokeSuspend执行
               if (var10000 == var6) {
                  return var6;
               }
               break;
            case 1:
               aaa = (String)this.L$0;
               ResultKt.throwOnFailure($result);
               var10000 = $result;
               break;
            default:
               throw new IllegalStateException("call to 'resume' before 'invoke' with coroutine");
            }

            int bbb = ((Number)var10000).intValue();
            String var4 = aaa + bbb;
            boolean var5 = false;
            System.out.print(var4);
            return Unit.INSTANCE;
         }

         @NotNull
         public final Continuation create(@Nullable Object value, @NotNull Continuation completion) {
            Intrinsics.checkNotNullParameter(completion, "completion");
            Function2 var3 = new <anonymous constructor>(completion);
            return var3;
         }

         public final Object invoke(Object var1, Object var2) {
            return ((<undefinedtype>)this.create(var1, (Continuation)var2)).invokeSuspend(Unit.INSTANCE);
         }
      }), 2, (Object)null);
   }

   // $FF: synthetic method
   public static void main(String[] var0) {
      main();
   }
}

```



```java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        // 枚举 a
        for (int first = 0; first < n; ++first) {
            // 需要和上一次枚举的数不相同
            if (first > 0 && nums[first] == nums[first - 1]) {
                continue;
            }
            // c 对应的指针初始指向数组的最右端
            int third = n - 1;
            int target = -nums[first];
            // 枚举 b
            for (int second = first + 1; second < n; ++second) {
                // 需要和上一次枚举的数不相同
                if (second > first + 1 && nums[second] == nums[second - 1]) {
                    continue;
                }
                // 需要保证 b 的指针在 c 的指针的左侧
                while (second < third && nums[second] + nums[third] > target) {
                    --third;
                }
                // 如果指针重合，随着 b 后续的增加
                // 就不会有满足 a+b+c=0 并且 b<c 的 c 了，可以退出循环
                if (second == third) {
                    break;
                }
                if (nums[second] + nums[third] == target) {
                    List<Integer> list = new ArrayList<Integer>();
                    list.add(nums[first]);
                    list.add(nums[second]);
                    list.add(nums[third]);
                    ans.add(list);
                }
            }
        }
        return ans;
    }
}


```



channel热流，flow冷流

http://events.jianshu.io/p/b0f1b82e95cf

#### 基本使用

1、创建的方式

```kotlin
fun start(){
  //方式1: 启动一个新的携程，会阻塞当前线程，直到里面的代码执行完
  runBlocking{
  
   }
  //方式2: 在全局作用域，开启子线程，不会阻塞主线程，返回job
  GlobalScope.launch{
    
  }
  
  GlobalScope.async{
    
  }
}
```

2、coroutineContext 协程上下文

> - 线程行为、生命周期、异常以及提调试
> - 包含用户定义的一些数据集合，这些数据与协程密切相关
> - 它是一个有索引的Element实例集合，一个介于set和map之间的数据结构，每个element在这个集合有一个唯一的key
>
> -- Job: 控制协程的生命周期
>
> -- coroutineDispatcher：向合适的线程分发任务
>
> -- coroutineName: 协程的名称，吊饰的时候很有用
>
> -- coroutineExecptionHandler: 处理未补获的异常

3、 CoroutineDispatcher：

> -  Dispatchers.Default
>
>   默认的调度器，适合处理后台计算，是一个cpu密集型调度器。如果创建coroutine的时候没有制定dispatcher，则一般默认使用这个作为默认值。Default dispatcher 使用一个共享的后台线程池来运行里面的任务。注意它和i o共享线程池，只不过限制了最大并发次数不同。
>
> - Dispatchers.IO
>
>   顾名思义这个是用来执行阻塞i o操作的，是和Default公用一个共享的线程池来执行里面的任务，根据同时运行的任务数量，在需要的时候会创建额外的线程，当任务执行完毕后会释放不需要的线程。
>
> - Dispatchers.Unconfined
>
>   由于ispatchers.Unconfined未定义线程池，所以执行的时候默认在启动线程。遇到第一个挂起点，之后调用resume的线程决定恢复协程的线程。
>
> - Dispatchers.Main
>
>   指定执行的线程是主线程，在Android上就是UI线程
>
> 由于子coroutine会继承父coroutine的context，所以为了方便使用，我们一般会在父coroutine上设置一个dispatcher，然后所有子coroutine自动使用这个dispatcher。

4、coroutineStart:

> - coroutineStart.DEFAULT:
>
>    协程创建后立即开启调度，在调度前如果协程被取消，其将直接进入取消响应的状态，但也有可能在执行前被取消。
>
> - coroutineStart.ATOMIC
>
>   协程创建后立即开始调度，协程执行到第一个挂起点之前不响应取消，虽然是立即调度，但其将调度和执行的两个步骤合二为一，就像她的名字一样，其保证调度和执行是原子操作，因此协程也一定会执行
>
> - coroutineStart.LAZY
>
>   只要协程被需要时，包括主动调用该协程的start、join或者await等函数时才会开始调度，如果调度前被取消，协程将直接进入异常结束状态
>
> - coroutineStart.UNDISPATCHED
>
>   协程创建后立即在当前函数调用栈中执行，知道遇到第一个真正挂起的点事立即执行，因此协程一定会执行。

4、coroutineScope协程作用域: 

> 只是定义了一个新的coroutine的执行scope。每个coroutine builder都是coroutineScope的口站函数，并且自动的继承了当前scope的coroutineContex
>
> 官方框架在实现复合协程的过程中也提供了作用域，主要用于明确协程之间的父子关系，以及对于取消或者异常处理等方面的传播行为。
>
> 该作用域包括以下三种：
>
> - 顶级作用域
>
>   没有父协程的携程所在的作用域为顶级
>
> - 协同作用域
>
>   携程中启动新的携程，新的携程为所在携程的子携程，这种情况下，子携程所在的作用域默认为协同作用域。此时子携程抛出的未捕捉的异常，都会传递给父携程处理，父携程同时也会被取消。
>
>   ```kotlin
>   //croutineScope内部的异常会向上传播，子携程未捕获的异常会向上传递给父携程，任何一个子携程异常退出，会导致整体退出
>   public suspend fun <R> coroutineScope(block: suspend CoroutineScope.()->R): R{
>     contract{
>       callsInPlace(block, InvocationKind.EXACTLY_ONCE)
>     }
>     return suspendCoroutineUninterceptedReturn{
>       uCont->
>       val coroutine = ScopeCoroutine(uCont.context, uCoint)
>       cotoutine.startUndispatchedOrReturn(coroutine, block)
>     }
>   }
>   ```
>
>   
>
> - 主从作用域
>
>   与协同作用域在携程的父子关系上一致，区别在于，处于该作用域下的携程出现未捕获的异常时，不会将异常向上传递给父携程。
>
>   suspervisorScope属于主从作用域，会继承父携程的上下文，它的特点就是子携程的异常不会影响父携程
>
>   ```kotlin
>   public suspend fun <R> suspervisorScope(block: suspend CoroutineScope.()->R): R{
>     contract{
>       callsInPlace(block, InvocationKind.EXACTLY_ONCE)
>     }
>     return suspendCoroutineUninterceptedReturn{
>       uCont->
>       val coroutine = ScopeCoroutine(uCont.context, uCoint)
>       cotoutine.startUndispatchedOrReturn(coroutine, block)
>     }
>   }
>   ```
>
>   

#### 异常

java: 受检查的异常  kotlin：没有

kotlin携程异常处理

从返回值来看：

1. job: scope.launch{}

   > Try{}catch{}、CoroutineExceptionHandler  都有效
   >
   > ```kotlin
   > val handler = CoroutineExceptionHandler{
   >   _,e->
   >   log("e=${e.stackTraceToString}")
   > }
   > 
   > scope.launch(handler){}
   > ```
   >
   > 
   >
   > 协同：设置在根作用域有效，子携程发送了异常，所有的携程都会取消掉

2. Defferred: scope.async{}

   需要在job.await捕获，其他方法都不行

   > val job = scope.async{}
   >
   > Try{
   >
   > job.await}catch(e){
   >
   > }

全局异常处理(只对launch)：

> ```kotlin
> //1.定义全局处理类
> class GlobleHandler: CoroutineExecptionHandler{
>   override val key: CoroutineContext.Key<*> 
>   get() = CoroutineExecptionHandler
>   
>   //ServerLoader SPI
>   override fun handleExecption(context: CoroutineContext, execption: Throwable){
>     
>     //这里只会响应，要想不挂，还是得setDefaultUncaughtExexption捕获，因为底层还有个AndroidExexptionHandle处理，如果没有setDefaultUncaughtExexption会抛出崩溃
>     log("e=${e.stackTraceToString}")
>     Thread.setDefaultUncaughtExexption{
>     }
>   }
> }
> ```
>
> 需要在resource中配置META-INF.service配置GlobleHandler全路径，利用jv m的serviceloader

 总结：

1. [协同]携程中异常传播是双向的，子携程的异常会导致父携程结束，父携程的异常也会导致整个结束

2. 【主从】supervisorJob的异常时单向传播的，子携程的异常不影响其他，父携程的异常会影响全局，因此supervisorJob的子携程都应该设置捕获异常句柄

   //注意：亲儿子：单向传播，孙子：//协同

   ```kotlin
   val scope = MainScope() //主从
   scope.launch{ //亲儿子，协同
     launch{
       //孙子这里崩了会导致儿子崩溃
     }
     launch{
       
     }
   }
   ```

3. Job的异常需要在根携程域捕获，类似Thread.defaultUncaughtExexptionHandler，在子携程无法捕获，会奔溃

4. defferd的异常依赖用户来最终消费异常，如果没消费，不需要捕获，也不会崩溃，如果消费了，则需要捕获，不然会崩溃

#### 作用域和上下文

从不同的使用角度来看

- 作用域：管理携程

  > 比如协同作用域：发生了异常，异常如何传播，双向传播，一荣俱荣一损俱损
  >
  > 主从作用域：单向传播，父亲出了问题，儿子都晚了，儿子出事了，不影响父亲以及兄弟姐妹
  >
  > 
  >
  > 不像在activity中开启线程，什么时候干完，有没有出生命问题，线程中又开启线程，没有统一的出入口
  >
  > 
  >
  > 携程就保证了黑盒性质：一个入口，一个出口
  >
  > 结构化并发：满足这个黑盒性质。

- 上下文：记录了携程的运行环境【name,线程，job】，可以随时的去配置和获取，跟activity中的context一样，获取资源